File: python/vedadb/migrations.py
"""
Schema migration runner for VedaDB.

Manages database schema migrations with versioning, rollback support,
and dependency tracking.  Migrations are defined as Python classes with
``upgrade`` and ``downgrade`` methods.

Example — define a migration::

    from vedadb.migrations import Migration, MigrationRunner

    class AddUsersTable(Migration):
        version = 1
        name = "add_users_table"
        dependencies = []

        def upgrade(self, db):
            db.query('''
                CREATE TABLE users (
                    id INTEGER PRIMARY KEY,
                    name TEXT NOT NULL,
                    email TEXT UNIQUE
                );
            ''')

        def downgrade(self, db):
            db.query("DROP TABLE users;")

    runner = MigrationRunner(db, migrations=[AddUsersTable])
    runner.migrate()  # Runs all pending migrations
"""

from __future__ import annotations

import logging
import time
from dataclasses import dataclass, field
from typing import Any, Callable, ClassVar, Dict, List, Optional, Type

from .exceptions import VedaDBQueryError, VedaDBValidationError

logger = logging.getLogger("vedadb.migrations")


# ---------------------------------------------------------------------------
# Migration base class
# ---------------------------------------------------------------------------

class Migration:
    """Base class for schema migrations.

    Subclasses must define ``version``, ``name``, and implement
    ``upgrade`` and ``downgrade`` methods.

    Attributes:
        version: Integer version number (must be unique, monotonic).
        name: Human-readable migration name.
        dependencies: List of version numbers this migration depends on.
    """

    version: ClassVar[int] = 0
    name: ClassVar[str] = ""
    dependencies: ClassVar[list[int]] = []

    def upgrade(self, db: Any) -> None:
        """Apply this migration. Must be implemented by subclasses."""
        raise NotImplementedError

    def downgrade(self, db: Any) -> None:
        """Revert this migration. Must be implemented by subclasses."""
        raise NotImplementedError

    @classmethod
    def describe(cls) -> str:
        return f"{cls.version}: {cls.name}"

    def __repr__(self) -> str:
        return f"<Migration v{self.version} {self.name!r}>"


# ---------------------------------------------------------------------------
# Migration record
# ---------------------------------------------------------------------------

@dataclass
class MigrationRecord:
    """A record of an applied migration in the database."""

    version: int
    name: str
    applied_at: float
    checksum: str = ""

    def to_row(self) -> dict:
        return {
            "version": self.version,
            "name": self.name,
            "applied_at": self.applied_at,
            "checksum": self.checksum,
        }


# ---------------------------------------------------------------------------
# MigrationRunner
# ---------------------------------------------------------------------------

class MigrationRunner:
    """Schema migration runner.

    Manages the migration lifecycle: tracks which migrations have been
    applied, runs pending ones in order, and supports rollback.

    Args:
        db: A :class:`VedaDB` or compatible client.
        migrations: List of :class:`Migration` subclasses to manage.
        table_name: Name of the migrations tracking table.
        auto_create_table: Whether to auto-create the tracking table.

    Example::

        runner = MigrationRunner(db, migrations=[
            AddUsersTable,
            AddPostsTable,
        ])
        runner.migrate()      # Apply all pending
        runner.rollback(1)    # Rollback one migration
        runner.status()       # Show current state
    """

    def __init__(
        self,
        db: Any,
        migrations: list[Type[Migration]] | None = None,
        table_name: str = "_migrations",
        auto_create_table: bool = True,
    ):
        self._db = db
        self._migrations: list[Type[Migration]] = sorted(
            (migrations or []), key=lambda m: m.version
        )
        self._table_name = table_name
        self._auto_create = auto_create_table

        if auto_create_table:
            self._ensure_migration_table()

    # ------------------------------------------------------------------
    # Core migration logic
    # ------------------------------------------------------------------

    def migrate(self, target_version: int | None = None) -> list[MigrationRecord]:
        """Apply all pending migrations (or up to *target_version*).

        Args:
            target_version: If provided, stop at this version.

        Returns:
            List of :class:`MigrationRecord` for applied migrations.
        """
        applied = self._get_applied_versions()
        records: list[MigrationRecord] = []

        for migration_class in self._migrations:
            if migration_class.version in applied:
                continue
            if target_version is not None and migration_class.version > target_version:
                break

            # Check dependencies
            for dep in migration_class.dependencies:
                if dep not in applied:
                    raise VedaDBValidationError(
                        f"Migration {migration_class.version} depends on {dep} "
                        f"which has not been applied"
                    )

            # Apply
            instance = migration_class()
            logger.info("Applying migration %s", instance.describe())
            start = time.perf_counter()
            instance.upgrade(self._db)
            elapsed = (time.perf_counter() - start) * 1000.0

            record = MigrationRecord(
                version=migration_class.version,
                name=migration_class.name,
                applied_at=time.time(),
            )
            self._record_migration(record)
            records.append(record)
            logger.info(
                "Migration %s applied in %.1fms", instance.describe(), elapsed
            )

        if not records:
            logger.info("No pending migrations")
        else:
            logger.info("Applied %d migration(s)", len(records))

        return records

    def rollback(self, steps: int = 1) -> list[MigrationRecord]:
        """Rollback the last *steps* migrations.

        Args:
            steps: Number of migrations to rollback.

        Returns:
            List of :class:`MigrationRecord` for rolled back migrations.
        """
        applied = self._get_applied_migrations()
        to_rollback = list(reversed(applied))[:steps]
        records: list[MigrationRecord] = []

        for record in to_rollback:
            migration_class = self._find_migration(record.version)
            if migration_class is None:
                logger.warning("Migration %d not found, skipping rollback", record.version)
                continue

            instance = migration_class()
            logger.info("Rolling back migration %s", instance.describe())
            instance.downgrade(self._db)
            self._remove_migration(record.version)
            records.append(record)
            logger.info("Rolled back migration %s", instance.describe())

        return records

    def rollback_to(self, target_version: int) -> list[MigrationRecord]:
        """Rollback all migrations down to (but not including) *target_version*.

        Args:
            target_version: Target version to rollback to.

        Returns:
            List of rolled back :class:`MigrationRecord`.
        """
        applied = self._get_applied_migrations()
        to_rollback = [r for r in reversed(applied) if r.version > target_version]
        records: list[MigrationRecord] = []

        for record in to_rollback:
            migration_class = self._find_migration(record.version)
            if migration_class is None:
                continue
            instance = migration_class()
            logger.info("Rolling back migration %s", instance.describe())
            instance.downgrade(self._db)
            self._remove_migration(record.version)
            records.append(record)

        return records

    # ------------------------------------------------------------------
    # Status and inspection
    # ------------------------------------------------------------------

    def status(self) -> dict:
        """Return current migration status.

        Returns:
            Dict with ``current_version``, ``pending``, and ``applied``.
        """
        applied_versions = self._get_applied_versions()
        all_versions = {m.version for m in self._migrations}
        pending = sorted(all_versions - applied_versions)

        return {
            "current_version": max(applied_versions) if applied_versions else 0,
            "applied_count": len(applied_versions),
            "pending_count": len(pending),
            "pending": pending,
            "applied": sorted(applied_versions),
        }

    def pending(self) -> list[Type[Migration]]:
        """Return list of pending migration classes."""
        applied = self._get_applied_versions()
        return [m for m in self._migrations if m.version not in applied]

    def applied(self) -> list[MigrationRecord]:
        """Return list of applied migration records."""
        return self._get_applied_migrations()

    def is_at_latest(self) -> bool:
        """Return True if all migrations have been applied."""
        return len(self.pending()) == 0

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _ensure_migration_table(self) -> None:
        """Create the migrations tracking table if it doesn't exist."""
        sql = f"""
            CREATE TABLE IF NOT EXISTS {self._table_name} (
                version INTEGER PRIMARY KEY,
                name TEXT NOT NULL,
                applied_at REAL NOT NULL,
                checksum TEXT
            );
        """
        try:
            self._db.query(sql)
        except VedaDBQueryError:
            pass

    def _get_applied_versions(self) -> set[int]:
        """Return the set of already-applied migration versions."""
        try:
            result = self._db.query(f"SELECT version FROM {self._table_name};")
            if result and result.rows:
                return {int(row[0]) for row in result.rows}
        except VedaDBQueryError:
            pass
        return set()

    def _get_applied_migrations(self) -> list[MigrationRecord]:
        """Return ordered list of applied migration records."""
        try:
            result = self._db.query(
                f"SELECT version, name, applied_at, checksum "
                f"FROM {self._table_name} ORDER BY version DESC;"
            )
            if result and result.rows:
                return [
                    MigrationRecord(
                        version=int(row[0]),
                        name=str(row[1]),
                        applied_at=float(row[2]),
                        checksum=str(row[3]) if len(row) > 3 else "",
                    )
                    for row in result.rows
                ]
        except VedaDBQueryError:
            pass
        return []

    def _record_migration(self, record: MigrationRecord) -> None:
        """Record a migration as applied."""
        from .protocol import sql_literal
        sql = (
            f"INSERT INTO {self._table_name} "
            f"(version, name, applied_at, checksum) "
            f"VALUES ({record.version}, {sql_literal(record.name)}, "
            f"{record.applied_at}, {sql_literal(record.checksum)});"
        )
        self._db.query(sql)

    def _remove_migration(self, version: int) -> None:
        """Remove a migration record (for rollback)."""
        self._db.query(f"DELETE FROM {self._table_name} WHERE version = {version};")

    def _find_migration(self, version: int) -> Type[Migration] | None:
        """Find a migration class by version."""
        for m in self._migrations:
            if m.version == version:
                return m
        return None

    def __repr__(self) -> str:
        status = self.status()
        return (
            f"<MigrationRunner current=v{status['current_version']} "
            f"pending={status['pending_count']}>"
        )


# ---------------------------------------------------------------------------
# Decorator for inline migrations
# ---------------------------------------------------------------------------


def migration(version: int, name: str | None = None, dependencies: list[int] | None = None):
    """Decorator to register a function as a migration.

    Example::

        @migration(version=1, name="create_users")
        def create_users(db):
            db.query("CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT);")

        @migration(version=2, name="add_email", dependencies=[1])
        def add_email(db):
            db.query("ALTER TABLE users ADD COLUMN email TEXT;")
    """
    def decorator(fn: Callable[[Any], None]) -> Type[Migration]:
        class_name = f"Migration_{version}_{fn.__name__}"
        attrs = {
            "version": version,
            "name": name or fn.__name__,
            "dependencies": dependencies or [],
            "upgrade": lambda self, db: fn(db),
            "downgrade": lambda self, db: None,
        }
        return type(class_name, (Migration,), attrs)

    return decorator

File: node/src/orm/migrations.js
/**
 * VedaDB ORM Migration Runner
 *
 * Manages database schema migrations with up/down support.
 *
 * @example
 * const { MigrationRunner } = require('vedadb/orm/migrations');
 *
 * const runner = new MigrationRunner(client, './migrations');
 * await runner.up();        // Run all pending migrations
 * await runner.down(1);     // Rollback 1 migration
 * await runner.status();    // Show migration status
 */

'use strict';

const fs = require('fs');
const path = require('path');

/**
 * Represents a single migration.
 */
class Migration {
  constructor(version, name, upFn, downFn) {
    this.version = version;
    this.name = name;
    this.up = upFn;
    this.down = downFn;
  }
}

/**
 * Migration runner that manages schema changes.
 */
class MigrationRunner {
  constructor(client, migrationsDir) {
    this.client = client;
    this.migrationsDir = migrationsDir;
    this.migrations = [];
  }

  /**
   * Initialize the migrations tracking table.
   */
  async init() {
    await this.client.execute(`
      CREATE TABLE IF NOT EXISTS _schema_migrations (
        version TEXT PRIMARY KEY,
        name TEXT NOT NULL,
        applied_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
      );
    `);
  }

  /**
   * Load migrations from the migrations directory.
   */
  loadMigrations() {
    if (!fs.existsSync(this.migrationsDir)) {
      return;
    }
    const files = fs.readdirSync(this.migrationsDir)
      .filter(f => f.endsWith('.js'))
      .sort();

    for (const file of files) {
      const filepath = path.join(this.migrationsDir, file);
      const mod = require(filepath);
      const version = file.split('_')[0];
      const name = file.replace(/\.js$/, '');
      this.migrations.push(new Migration(
        version,
        name,
        mod.up || (async () => {}),
        mod.down || (async () => {})
      ));
    }
  }

  /**
   * Get list of already applied migrations.
   */
  async appliedMigrations() {
    await this.init();
    const result = await this.client.query(
      'SELECT version FROM _schema_migrations ORDER BY version;'
    );
    return new Set(result.rows.map(r => r[0]));
  }

  /**
   * Run all pending migrations.
   */
  async up() {
    this.loadMigrations();
    const applied = await this.appliedMigrations();
    const pending = this.migrations.filter(m => !applied.has(m.version));

    for (const migration of pending) {
      console.log(`Applying migration: ${migration.name}`);
      await this.client.begin();
      try {
        await migration.up(this.client);
        await this.client.query(
          'INSERT INTO _schema_migrations (version, name) VALUES ($1, $2);',
          [migration.version, migration.name]
        );
        await this.client.commit();
        console.log(`Applied: ${migration.name}`);
      } catch (err) {
        await this.client.rollback();
        throw new Error(`Migration ${migration.name} failed: ${err.message}`);
      }
    }
    return pending.length;
  }

  /**
   * Rollback the last N migrations.
   */
  async down(count = 1) {
    this.loadMigrations();
    const result = await this.client.query(
      'SELECT version, name FROM _schema_migrations ORDER BY version DESC LIMIT $1;',
      [count]
    );
    const toRollback = result.rows.map(r => ({
      version: r[0],
      name: r[1],
    }));

    for (const row of toRollback) {
      const migration = this.migrations.find(m => m.version === row.version);
      if (!migration) {
        console.warn(`Migration ${row.version} not found for rollback`);
        continue;
      }
      console.log(`Rolling back: ${migration.name}`);
      await this.client.begin();
      try {
        await migration.down(this.client);
        await this.client.query(
          'DELETE FROM _schema_migrations WHERE version = $1;',
          [migration.version]
        );
        await this.client.commit();
        console.log(`Rolled back: ${migration.name}`);
      } catch (err) {
        await this.client.rollback();
        throw new Error(`Rollback ${migration.name} failed: ${err.message}`);
      }
    }
    return toRollback.length;
  }

  /**
   * Show migration status.
   */
  async status() {
    this.loadMigrations();
    const applied = await this.appliedMigrations();
    return this.migrations.map(m => ({
      version: m.version,
      name: m.name,
      applied: applied.has(m.version),
    }));
  }

  /**
   * Create a new migration file.
   */
  createMigration(name) {
    if (!fs.existsSync(this.migrationsDir)) {
      fs.mkdirSync(this.migrationsDir, { recursive: true });
    }
    const timestamp = new Date().toISOString().replace(/[^0-9]/g, '').slice(0, 14);
    const filename = `${timestamp}_${name}.js`;
    const filepath = path.join(this.migrationsDir, filename);

    const template = `'use strict';

/**
 * Migration: ${name}
 */

module.exports = {
  async up(client) {
    // Write your "up" migration here
    // Example:
    // await client.execute(
    //   'CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT NOT NULL);'
    // );
  },

  async down(client) {
    // Write your "down" migration here
    // Example:
    // await client.execute('DROP TABLE IF EXISTS users;');
  },
};
`;
    fs.writeFileSync(filepath, template);
    return filepath;
  }
}

module.exports = { MigrationRunner, Migration };
